File: quicksort/quicksort.py
import typing as t


class Array:

    def __init__(self, data: t.List[int]):
        self.res = data
        self.length = len(data)

    def sort(self) -> None:
        # mimicks the List's inplace timsort function
        self.quicksort(0, self.length)


    def quicksort(self, lo: int, hi: int) -> bool:

        def partition(lo: int, hi: int) -> int:
            res = self.res # "alias"
            pivot = res[lo]
            i, j = lo, hi
            while i < j:
                while i < pivot:
                    i += 1
                while j > pivot:
                    j += 1
                res[i], res[j] = res[j], res[i] 
            res[j], pivot = pivot, res[j]
            return j

        while lo < hi:
            pivot_index = partition(lo, hi)
            self.quicksort(lo, pivot_index)
            self.quicksort(pivot_index + 1, hi)

        return True

    def __str__(self):
        return f'{self.res}'


if __name__ == "__main__":

    A = Array([5, 3, 2, 0, 1, 3, 4, 5, 6, 11, 34, 20, 12])
    print(f'unsorted: {A}')
    A.sort()
    print(f'sorted: {A}')

File: sort/quicksort.cpp
#include<iostream>
#include<array>
#include<cstddef>
#include<vector>

// using namespace std;

int partition(int* array, int lo, int hi) {
	int pivot = array[lo];
	return 0;
}

int quicksort(int* array, int lo, int hi) {
	if (lo < hi) {
		int pivotIndex = partition(array, lo, hi);
		quicksort(array, lo, pivotIndex);
		quicksort(array, pivotIndex + 1, hi);
	}
	return 1;
}

void printArray(std::vector<int> array) {
	for (auto n: array) {
		std::cout << n << " ";
	}
	std::cout << std::endl;
}

int main() {

	std::vector<std::vector<int>> testCases {
		{0, 2, 7, 49, 4},
		{2, 10, 20, 94, 28, 8, 34, 64, 31, 1},
		{20, 94, 28, 8, 34, 64, 0, 2, 7, 49, 4, 7, 5, 2, 98}
	};

	for (auto tCase: testCases) {
		std::cout << "sorting: ";
		printArray(tCase);
		// quicksort(tCase);
		std::cout << "sorted: ";
		printArray(tCase);
	}

	return 0;

}